File: backend/app/models/payout_request.py
from datetime import datetime, timezone
from app.db.database import db

class PayoutRequest(db.Model):
    __tablename__ = "payout_requests"

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, nullable=False, index=True)

    # En tu BD estos son enums; indicamos el nombre para que SQLAlchemy use el existente
    account_type = db.Column(
        db.Enum("bank", "nequi", "daviplata", "bancolombia_cell", "other",
                name="account_type", create_type=False),
        nullable=False,
    )
    account_kind = db.Column(
        db.Enum("savings", "checking", name="account_kind", create_type=False),
        nullable=True,
    )

    # Banco (solo si account_type == 'bank')
    bank_id = db.Column(
        db.Integer,
        db.ForeignKey("public.banks.id"),  # si usas schema 'public'
        nullable=True,
        index=True,
)

    account_number = db.Column(db.String(50), nullable=False)

    # ← estos dos campos faltaban en tu modelo
    amount_micros  = db.Column(db.BigInteger, nullable=False, default=0)
    currency_code  = db.Column(db.String(3), nullable=False, default="COP")

    status = db.Column(
        db.Enum("requested", "processing", "paid", "rejected",
                name="payout_status", create_type=False),
        nullable=False,
        default="requested",
    )

    observations = db.Column(db.Text, nullable=True)

    requested_at = db.Column(
        db.DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    created_at = db.Column(
        db.DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at = db.Column(
        db.DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    def to_dict(self):
        return {
            "id": self.id,
            "user_id": self.user_id,
            "account_type": self.account_type,
            "account_kind": self.account_kind,
            "bank_id": self.bank_id,
            "account_number": self.account_number,
            "amount_micros": self.amount_micros,
            "currency_code": self.currency_code,
            "status": self.status,
            "observations": self.observations,
            "requested_at": self.requested_at.isoformat() if self.requested_at else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: lib/domain/models/payout_request.dart
// lib/domain/models/payout_request.dart

class PayoutRequestInput {
  /// 'bank' | 'nequi' | 'daviplata' | 'bancolombia_cell' | 'other'
  final String accountType;

  /// 'savings' | 'checking' (solo si accountType == 'bank')
  final String? accountKind;

  /// Número de cuenta o celular (según accountType)
  final String accountNumber;

  /// Código del banco (si accountType == 'bank')
  final String? bankCode;

  /// Observaciones opcionales que envía el usuario
  final String? observations;

  PayoutRequestInput({
    required this.accountType,
    this.accountKind,
    required this.accountNumber,
    this.bankCode,
    this.observations,
  })  : assert(
          accountNumber.trim().isNotEmpty,
          'accountNumber no puede estar vacío',
        ),
        assert(
          accountType.toLowerCase() != 'bank' ||
              (bankCode != null &&
                  bankCode.trim().isNotEmpty &&
                  accountKind != null &&
                  accountKind.trim().isNotEmpty),
          "Para accountType=bank debes enviar bankCode y accountKind ('savings'|'checking')",
        ),
        assert(
          accountType.toLowerCase() == 'bank' ||
              (bankCode == null && accountKind == null),
          "Para accountType distinto de 'bank' no envíes bankCode ni accountKind",
        ),
        assert(
          const {
            'bank',
            'nequi',
            'daviplata',
            'bancolombia_cell',
            'other',
          }.contains(accountType.toLowerCase()),
          "accountType inválido",
        );

  Map<String, dynamic> toJson() => {
        'account_type': accountType.toLowerCase().trim(),
        if (accountKind != null) 'account_kind': accountKind!.toLowerCase().trim(),
        'account_number': accountNumber.trim(),
        if (bankCode != null) 'bank_code': bankCode!.trim(),
        if (observations != null && observations!.isNotEmpty)
          'observations': observations,
      };

  @override
  String toString() =>
      'PayoutRequestInput('
      'type=$accountType, '
      'kind=$accountKind, '
      'number=$accountNumber, '
      'bank=$bankCode, '
      'obs=$observations'
      ')';
}

/// Modelo de respuesta al crear/consultar una solicitud de retiro.
class PayoutRequest {
  final int id;
  final int amountMicros;
  final String currencyCode; // p. ej. 'COP'
  final String status;       // p. ej. 'pending', 'processing', 'paid', 'rejected'
  final DateTime? requestedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const PayoutRequest({
    required this.id,
    required this.amountMicros,
    required this.currencyCode,
    required this.status,
    this.requestedAt,
    this.createdAt,
    this.updatedAt,
  });

  /// Conveniencia: monto en COP (si tu moneda base es COP).
  double get amountCop => amountMicros / 1_000_000.0;

  factory PayoutRequest.fromJson(Map<String, dynamic> j) {
    int asInt(dynamic v) =>
        v is int ? v : (v is num ? v.toInt() : int.tryParse('$v') ?? 0);

    DateTime? asDate(dynamic v) {
      if (v == null) return null;
      if (v is DateTime) return v;
      return DateTime.tryParse(v.toString());
    }

    return PayoutRequest(
      id: asInt(j['id'] ?? j['request_id']),
      amountMicros: asInt(j['amount_micros'] ?? j['amountMicros']),
      currencyCode:
          (j['currency_code'] ?? j['currencyCode'] ?? 'COP').toString(),
      status: (j['status'] ?? '').toString(),
      requestedAt: asDate(j['requested_at'] ?? j['requestedAt']),
      createdAt: asDate(j['created_at'] ?? j['createdAt']),
      updatedAt: asDate(j['updated_at'] ?? j['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'amount_micros': amountMicros,
        'currency_code': currencyCode,
        'status': status,
        'requested_at': requestedAt?.toIso8601String(),
        'created_at': createdAt?.toIso8601String(),
        'updated_at': updatedAt?.toIso8601String(),
      };

  @override
  String toString() =>
      'PayoutRequest(id=$id, amountMicros=$amountMicros, '
      'currency=$currencyCode, status=$status, '
      'requestedAt=$requestedAt, createdAt=$createdAt, updatedAt=$updatedAt)';
}
